crates/circuit/src/packed_instruction.rs

```diff
@@ -39,132 +42,375 @@ enum PackedOperationType {
+        );
+        assert!(
+            ((1 << StandardInstructionBits::DISCRIMINANT_BITS) - 1)
+                << StandardInstructionBits::DISCRIMINANT_OFFSET
+                == Self::DISCRIMINANT_MASK,
+            "(StandardInstructionBits) discriminant MUST be the 3 lowest bits!"
+        );
+        assert!(
+            ((1 << PointerBits::DISCRIMINANT_BITS) - 1) << PointerBits::DISCRIMINANT_OFFSET
+                == Self::DISCRIMINANT_MASK,
+            "(PointerBits) discriminant MUST be the 3 lowest bits!"
+        );
+    };
 
-impl PackedOperation {
-    /// The bits representing the `PackedOperationType` discriminant.  This can be used to mask out
-    /// the discriminant, and defines the rest of the bit shifting.
-    const DISCRIMINANT_MASK: usize = 0b11;
-    /// The number of bits used to store the discriminant metadata.
-    const DISCRIMINANT_BITS: u32 = Self::DISCRIMINANT_MASK.count_ones();
-    /// A bitmask that masks out only the standard gate information.  This should always have the
-    /// same effect as `POINTER_MASK` because the high bits should be 0 for a `StandardGate`, but
-    /// this is defensive against us adding further metadata on `StandardGate` later.  After
-    /// masking, the resulting integer still needs shifting downwards to retrieve the standard gate.
-    const STANDARD_GATE_MASK: usize = (u8::MAX as usize) << Self::DISCRIMINANT_BITS;
-    /// A bitmask that retrieves the stored pointer directly.  The discriminant is stored in the
-    /// low pointer bits that are guaranteed to be 0 by alignment, so no shifting is required.
-    const POINTER_MASK: usize = usize::MAX ^ Self::DISCRIMINANT_MASK;
-
-    /// Extract the discriminant of the operation.
     #[inline]
     fn discriminant(&self) -> PackedOperationType {
-        ::bytemuck::checked::cast((self.0 & Self::DISCRIMINANT_MASK) as u8)
+        // SAFETY: we read (just!) the discriminant from any of the union's members,
+        // since we guarantee it is found in the same place for all bitfields.
+        unsafe { self.gate.discriminant() }
+    }
+}
+
+impl From<StandardGateBits> for BitField {
+    fn from(gate: StandardGateBits) -> Self {
+        Self { gate }
+    }
+}
+
+impl From<StandardInstructionBits> for BitField {
+    fn from(instruction: StandardInstructionBits) -> Self {
+        Self { instruction }
+    }
+}
+
+impl From<PointerBits> for BitField {
+    fn from(pointer: PointerBits) -> Self {
+        Self { pointer }
+    }
+}
+
+// Used by bitfield-struct to convert into and out of the bitfield.
+impl StandardGate {
+    const fn into_bits(self) -> u8 {
+        self as _
+    }
+    const fn from_bits(value: u8) -> Self {
+        if value as usize >= STANDARD_GATE_SIZE {
+            panic!("unexpected standard gate type!")
+        }
+        // It would've been nice to use bytemuck here, but it doesn't support const
+        // casting for enums.
+        unsafe { std::mem::transmute(value) }
+    }
+}
```

Comment:
I did this so they'd be private to this file. I agree that it would be safer for these to be defined beside `StandardGate`, like we do with the `bytemuck` stuff.

Reply:
If you do keep these and want them private to the file, can you make them stand-alone functions (not `impl`) and use the `from` and `into` macro attributes to do it instead? Probably just minor style, but it'd make it easier to parse for me.